pkg/serial/handshake.go

```diff
@@ -28,134 +47,97 @@ import (
 )
 
 const (
-	flagSyn      byte = 0x16
-	flagAck      byte = 0x06
-	flagDebugAck byte = 0x07
-	flagNak      byte = 0x15
+	flagSyn byte = 0x16
+	flagAck      = 0x06
+	flagNak      = 0x15
 )
 
 // HandshakeError should only occure if the protocol between HandshakeServer and HandshakeClient was violated.
 type HandshakeError struct {
 	msg string
 }
 
-func (he *HandshakeError) Error() string { return "HandshakeServer: " + he.msg }
+func (he *HandshakeError) Error() string {
+	return he.msg
+}
+func init() {
+	rand.Seed(time.Now().UTC().UnixNano())
+}
 
-// PurgeIncoming is used to clear a channel of bytes prior to handshaking
-func PurgeIncoming(conn net.Conn) {
-	if tracing {
-		defer trace.End(trace.Begin(""))
+// ReadAtLeastN reads at least l bytes and returns those l bytes or errors
+func ReadAtLeastN(conn io.ReadWriter, buffer []byte, l int) ([]byte, error) {
+	n, err := io.ReadAtLeast(conn, buffer, l)
+	if err != nil {
+		return nil, err
 	}
-	buf := make([]byte, 255)
 
-	// read until the incoming channel is empty
-	conn.SetReadDeadline(time.Now().Add(time.Duration(10 * time.Millisecond)))
-	for n, err := conn.Read(buf); n != 0 || err == nil; n, err = conn.Read(buf) {
-		log.Debugf("discarding following %d bytes from input channel\n", n)
-		log.Debugf("%+v\n", buf[0:n])
+	// however if we read more than l, it means buffer is not empty
+	if n != l {
+		buffer = buffer[n-l:]
```

Comment:
We get a lots of garbage data when we get the initial connection this handshake supposed to clear them and leave the connection in a known state so that the real ssh handshake can start. Also the client and server is looping with different frequencies (their read timeouts are 10ms vs 100ms I believe) so client could send multiple requests because of different timeouts. By getting the last l bytes we are saying that we are not interested with garbage data and also eliminating duplicated flags by only using the last one